kwok/cloudprovider/helpers.go

```diff
@@ -17,104 +17,131 @@ limitations under the License.
+	awsRegexp = regexp.MustCompile(`^\w+\.(nano|micro|small|medium|large|\d*xlarge|metal)$`)
+
+	familyDelim = regexp.MustCompile(`[.-]`)
+)
+
 type InstanceTypeOptions struct {
-	Name               string
-	Offerings          cloudprovider.Offerings
-	Architecture       string
-	OperatingSystems   sets.Set[string]
-	Resources          v1.ResourceList
-	InstanceTypeLabels map[string]string
+	Name             string                  `json:"name"`
+	Offerings        cloudprovider.Offerings `json:"offerings"`
+	Architecture     string                  `json:"architecture"`
+	OperatingSystems []v1.OSName             `json:"operatingSystems"`
+	Resources        v1.ResourceList         `json:"resources"`
+
+	// These are used for setting default requirements, they should not be used
+	// for setting arbitrary node labels.  Set the labels on the created NodePool for
+	// that use case.
+	instanceTypeLabels map[string]string
+}
+
+//go:embed instance_types.json
+var rawInstanceTypes []byte
+
+// ConstructInstanceTypes create many instance types based on the embedded instance type data
+func ConstructInstanceTypes() ([]*cloudprovider.InstanceType, error) {
+	var instanceTypes []*cloudprovider.InstanceType
+	var instanceTypeOptions []InstanceTypeOptions
+
+	if err := json.Unmarshal(rawInstanceTypes, &instanceTypeOptions); err != nil {
+		return nil, fmt.Errorf("could not parse JSON data: %w", err)
+	}
+
+	for _, opts := range instanceTypeOptions {
+		opts = setDefaultOptions(opts)
+		instanceTypes = append(instanceTypes, newInstanceType(opts))
+	}
+	return instanceTypes, nil
 }
 
-func MakeInstanceTypeLabels(cpu, memFactor int) map[string]string {
-	size := fmt.Sprintf("%dx", cpu)
-	var family string
-	switch memFactor {
-	case 2:
-		family = "c" // cpu
-	case 4:
-		family = "s" // standard
-	case 8:
-		family = "m" // memory
-	default:
-		family = "e" // exotic
+// parseSizeFromType will attempt to discover the instance size if it matches a special AWS format.
+// If not, fall back to the cpu value.  This works for both Azure and GCP (and the "generic" instances
+// generated by tools/gen_instances.go)
+func parseSizeFromType(ty, cpu string) string {
+	if matches := awsRegexp.FindStringSubmatch(ty); matches != nil {
+		return matches[1]
 	}
-	return map[string]string{
-		InstanceSizeLabelKey:   size,
-		InstanceFamilyLabelKey: family,
-		InstanceCPULabelKey:    fmt.Sprintf("%d", cpu),
-		InstanceMemoryLabelKey: fmt.Sprintf("%d", cpu*memFactor*1024),
+
+	return cpu
+}
+
+// parseFamilyFromType will attempt to discover the instance family from the node type.  The three major cloud providers
+// have the following formats:
+//
+// AWS   - [Family].[Size]
+// GCP   - [Family]-[Configuration]-[# of vCPUs])
+// Azure - [Family] + [Sub-family]* + [# of vCPUs] + ... + [Version]
+//
+// So here, we split on [.-], and if that fails, fall back to the first character of the instance type name.
+func parseFamilyFromType(instanceType string) string {
```

Comment:
Since this user-input isn't validated, is it worth doing a quick length check so that we're not going to out of bounds panic here, since L97 assumes its non-empty?